Extend IR to take customisable inputs for aesthetics

arj119 opened the pull request:
### Clubhouse Ticket
[ch132](https://app.clubhouse.io/manim-dsl/story/132/extend-ir-python-library-to-take-customisable-inputs)

### Description

Extended the Python classes `Rectangle`, `Code_block` and `Init_structure` to take color, and text properties. I then updated the Shape classes to take these properties. Implemented `StyleAttribute` and `Style` classes to handle and validate styles to only those which we support.

I have recently updated the python library to expose the constituent Manim objects that make up each shape. This is so that all or some can be modified in terms of colour etc. So in case of a rectangle you move the whole object with `rectangle.all` and can target just the text e.g. `rectangle.text.setColor(BLUE)`. The Kotlin Shape classes handle this nicely, with getters for the constituent objects and by default their `toString` methods return `$ident.all` referring to the whole VGroup in the object.

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update - potentially when style sheet is implemented

### How Has This Been Tested?

Implemented tests to see if python produced is correct and if validation occurs.

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

mayanksharma3 commented:
Haven't fully reviewed yet but just wanted to check, is there an option in this to set colors on an existing Mobject or shape? For example, when pushing I want the color to be red, but once it's on the stack I want it to be white

arj119 commented:
> Haven't fully reviewed yet but just wanted to check, is there an option in this to set colors on an existing Mobject or shape? For example, when pushing I want the color to be red, but once it's on the stack I want it to be white

Yes but you need references to the internal objects you want to set. You could do for example rectangle.text.set_color(BLUE) if text is exposed.

mayanksharma3 commented:
> > Haven't fully reviewed yet but just wanted to check, is there an option in this to set colors on an existing Mobject or shape? For example, when pushing I want the color to be red, but once it's on the stack I want it to be white
> 
> Yes but you need references to the internal objects you want to set. You could do for example rectangle.text.set_color(BLUE) if text is exposed.

Should we do that straight from the IR as manim code or should be abstract that inside the Rectangle class? We would ideally want to use FadeToColor (I think that's what it's called?) so it's a nice switch as a pose to a direct switch in color. That is also possible through Kotlin IR so it should be fine, but it could be added to the Rectangle class? Either works though as a solution

arj119 commented:
> > > Haven't fully reviewed yet but just wanted to check, is there an option in this to set colors on an existing Mobject or shape? For example, when pushing I want the color to be red, but once it's on the stack I want it to be white
> > 
> > 
> > Yes but you need references to the internal objects you want to set. You could do for example rectangle.text.set_color(BLUE) if text is exposed.
> 
> Should we do that straight from the IR as manim code or should be abstract that inside the Rectangle class? We would ideally want to use FadeToColor (I think that's what it's called?) so it's a nice switch as a pose to a direct switch in color. That is also possible through Kotlin IR so it should be fine, but it could be added to the Rectangle class? Either works though as a solution

Hm it really depends on how the references are handled. I feel like in our current implementation we cannot access the Text Object from the instantiated Rectangle as a reference is not stored anywhere.

iz0825 commented:
> > > > Haven't fully reviewed yet but just wanted to check, is there an option in this to set colors on an existing Mobject or shape? For example, when pushing I want the color to be red, but once it's on the stack I want it to be white
> > > 
> > > 
> > > Yes but you need references to the internal objects you want to set. You could do for example rectangle.text.set_color(BLUE) if text is exposed.
> > 
> > 
> > Should we do that straight from the IR as manim code or should be abstract that inside the Rectangle class? We would ideally want to use FadeToColor (I think that's what it's called?) so it's a nice switch as a pose to a direct switch in color. That is also possible through Kotlin IR so it should be fine, but it could be added to the Rectangle class? Either works though as a solution
> 
> Hm it really depends on how the references are handled. I feel like in our current implementation we cannot access the Text Object from the instantiated Rectangle as a reference is not stored anywhere.
> 
> EDIT: The above is true. A solution would be to change the init functions to construct and self assign the constituent MObjects and then on build we return the VGroup comprising them.

not sure if I understood what you were talking about correctly, but if we have the VGroup we have now, we can technically access the submobject if we know the index? If this is not what you were talking about, I'm sorry

arj119 commented:
> > > > > Haven't fully reviewed yet but just wanted to check, is there an option in this to set colors on an existing Mobject or shape? For example, when pushing I want the color to be red, but once it's on the stack I want it to be white
> > > > 
> > > > 
> > > > Yes but you need references to the internal objects you want to set. You could do for example rectangle.text.set_color(BLUE) if text is exposed.
> > > 
> > > 
> > > Should we do that straight from the IR as manim code or should be abstract that inside the Rectangle class? We would ideally want to use FadeToColor (I think that's what it's called?) so it's a nice switch as a pose to a direct switch in color. That is also possible through Kotlin IR so it should be fine, but it could be added to the Rectangle class? Either works though as a solution
> > 
> > 
> > Hm it really depends on how the references are handled. I feel like in our current implementation we cannot access the Text Object from the instantiated Rectangle as a reference is not stored anywhere.
> > EDIT: The above is true. A solution would be to change the init functions to construct and self assign the constituent MObjects and then on build we return the VGroup comprising them.
> 
> not sure if I understood what you were talking about correctly, but if we have the VGroup we have now, we can technically access the submobject if we know the index? If this is not what you were talking about, I'm sorry

You don't ever need to be sorry haha. Yes you're right about that but rather than keeping track of index we could just have a pointer so in `Rectangle.__init__()` we could have `self.text = Text(text, color...) `and so we can do rectangle1.text.fadeToColor(BLUE) or something similar.

iz0825 commented:
> > > > > > Haven't fully reviewed yet but just wanted to check, is there an option in this to set colors on an existing Mobject or shape? For example, when pushing I want the color to be red, but once it's on the stack I want it to be white
> > > > > 
> > > > > 
> > > > > Yes but you need references to the internal objects you want to set. You could do for example rectangle.text.set_color(BLUE) if text is exposed.
> > > > 
> > > > 
> > > > Should we do that straight from the IR as manim code or should be abstract that inside the Rectangle class? We would ideally want to use FadeToColor (I think that's what it's called?) so it's a nice switch as a pose to a direct switch in color. That is also possible through Kotlin IR so it should be fine, but it could be added to the Rectangle class? Either works though as a solution
> > > 
> > > 
> > > Hm it really depends on how the references are handled. I feel like in our current implementation we cannot access the Text Object from the instantiated Rectangle as a reference is not stored anywhere.
> > > EDIT: The above is true. A solution would be to change the init functions to construct and self assign the constituent MObjects and then on build we return the VGroup comprising them.
> > 
> > 
> > not sure if I understood what you were talking about correctly, but if we have the VGroup we have now, we can technically access the submobject if we know the index? If this is not what you were talking about, I'm sorry
> 
> You don't ever need to be sorry haha. Yes you're right about that but rather than keeping track of index we could just have a pointer so in `Rectangle.__init__()` we could have `self.text = Text(text, color...) `and so we can do rectangle1.text.fadeToColor(BLUE) or something similar.

ahh I see, yeah makes sense

arj119 commented:
> > > > > > > Haven't fully reviewed yet but just wanted to check, is there an option in this to set colors on an existing Mobject or shape? For example, when pushing I want the color to be red, but once it's on the stack I want it to be white
> > > > > > 
> > > > > > 
> > > > > > Yes but you need references to the internal objects you want to set. You could do for example rectangle.text.set_color(BLUE) if text is exposed.
> > > > > 
> > > > > 
> > > > > Should we do that straight from the IR as manim code or should be abstract that inside the Rectangle class? We would ideally want to use FadeToColor (I think that's what it's called?) so it's a nice switch as a pose to a direct switch in color. That is also possible through Kotlin IR so it should be fine, but it could be added to the Rectangle class? Either works though as a solution
> > > > 
> > > > 
> > > > Hm it really depends on how the references are handled. I feel like in our current implementation we cannot access the Text Object from the instantiated Rectangle as a reference is not stored anywhere.
> > > > EDIT: The above is true. A solution would be to change the init functions to construct and self assign the constituent MObjects and then on build we return the VGroup comprising them.
> > > 
> > > 
> > > not sure if I understood what you were talking about correctly, but if we have the VGroup we have now, we can technically access the submobject if we know the index? If this is not what you were talking about, I'm sorry
> > 
> > 
> > You don't ever need to be sorry haha. Yes you're right about that but rather than keeping track of index we could just have a pointer so in `Rectangle.__init__()` we could have `self.text = Text(text, color...) `and so we can do rectangle1.text.fadeToColor(BLUE) or something similar.
> 
> ahh I see, yeah makes sense

So it would be like on a separate line `self.play(rectangle.inside_text, GREEN))`. Since build returns the VGroup, we could probably do something like:

```
rectangle = Rectangle(text, color=...)
rectangle_vg = rectangle.build()
self.play(FadeToColor(rectangle.inside_text, GREEN))
```

So an extra reference is made to the Rectangle class and not just the VGroup the build returns.

arj119 commented:
> > > > > > > > Haven't fully reviewed yet but just wanted to check, is there an option in this to set colors on an existing Mobject or shape? For example, when pushing I want the color to be red, but once it's on the stack I want it to be white
> > > > > > > 
> > > > > > > 
> > > > > > > Yes but you need references to the internal objects you want to set. You could do for example rectangle.text.set_color(BLUE) if text is exposed.
> > > > > > 
> > > > > > 
> > > > > > Should we do that straight from the IR as manim code or should be abstract that inside the Rectangle class? We would ideally want to use FadeToColor (I think that's what it's called?) so it's a nice switch as a pose to a direct switch in color. That is also possible through Kotlin IR so it should be fine, but it could be added to the Rectangle class? Either works though as a solution
> > > > > 
> > > > > 
> > > > > Hm it really depends on how the references are handled. I feel like in our current implementation we cannot access the Text Object from the instantiated Rectangle as a reference is not stored anywhere.
> > > > > EDIT: The above is true. A solution would be to change the init functions to construct and self assign the constituent MObjects and then on build we return the VGroup comprising them.
> > > > 
> > > > 
> > > > not sure if I understood what you were talking about correctly, but if we have the VGroup we have now, we can technically access the submobject if we know the index? If this is not what you were talking about, I'm sorry
> > > 
> > > 
> > > You don't ever need to be sorry haha. Yes you're right about that but rather than keeping track of index we could just have a pointer so in `Rectangle.__init__()` we could have `self.text = Text(text, color...) `and so we can do rectangle1.text.fadeToColor(BLUE) or something similar.
> > 
> > 
> > ahh I see, yeah makes sense
> 
> So it would be like on a separate line `self.play(rectangle.inside_text, GREEN))`. Since build returns the VGroup, we could probably do something like:
> 
> ```
> rectangle = Rectangle(text, color=...)
> rectangle_vg = rectangle.build()
> self.play(FadeToColor(rectangle.inside_text, GREEN))
> ```
> 
> So an extra reference is made to the Rectangle class and not just the VGroup the build returns.

Addressed comments in latest commit

arj119 commented:
Closed the PR by accident whoops 😅

adityagoel4512 commented:
Code looks fine to me - unable to verify logic particularly well though so maybe worth waiting for someone else first!

mayanksharma3 commented on src/main/kotlin/com/manimdsl/shapes/Shapes.kt:
```diff
@@ -1,15 +1,108 @@
+        color?.let { style.addStyleAttribute(Color(color)) }
+        textColor?.let { style.addStyleAttribute(TextColor(textColor)) }
+        textWeight?.let { style.addStyleAttribute(TextWeight(textWeight)) }
+        font?.let { style.addStyleAttribute(Font(font)) }
```
Are all these let's necessary? You don't seem to use "it" inside the lambda?

arj119 replied:
Oh I can replace it with it :/

mayanksharma3 replied:
No but if it's working without using it can you not just call `style.addStyleAttribute(Color(color)) ` without the let?

mayanksharma3 replied:
@arj119

arj119 replied:
Oh no as if it is null it should not add anything

arj119 replied:
The defaults are in the python side not Kotlin. Otherwise the constructors generated in the main construct function will be messy

mayanksharma3 replied:
Ah okay cool looks good then
